File: javascript/state-of-tic-tac-toe/state-of-tic-tac-toe.js
//
// This is only a SKELETON file for the 'State of Tic Tac Toe' exercise. It's been provided as a
// convenience to get you started writing code faster.
//

/**
 * Compute the state of a tic-tac-toe board.
 * @param {Array[string]} board - Game board an array of three strings of length three. Space is a 
 * blank space on the board. Players are 'X', and 'O'
 * @returns {string} 'win' if X or O won, 'draw' if the game ended without a winner, 'ongoing' if the
 * game is incomplete. Throws errors if there is an invalid number of X or O characters (X always goes
 * first).
 */
export const gamestate = (board) => {
  
  let oCount = countPlayer(board, 'O')
  let xCount = countPlayer(board, 'X')
  if (oCount > xCount)
    throw new Error('Wrong turn order: O started')
  else if (xCount > oCount + 1)
    throw new Error('Wrong turn order: X went twice')

  let winX = isWon(board, 'X')
  let winO = isWon(board, 'O')
  let boardFull = isFull(board)

  if (winX && winO) 
  {
    throw new Error('Impossible board: game should have ended after the game was won')
  }

  if (!winX && !winO && boardFull)
    return 'draw'

  if (winX || winO)
    return 'win'

  return 'ongoing'

};

/**
 * Check if a tic-tac-toe board was won by the given player
 * @param {Array[string]} board - Array of three strings of length three representing the board
 * @param {string} player - The player to check for winning 'X', or 'O'
 * @returns (boolean) true if the player won, and false otherwise
 */
const isWon = (board, player) => {
  let current = board.map((row) => row.split(''))
  
  // rows & cols
  for(let i = 0; i < 3; i++)
  {
    if (current[i][0] === current[i][1] && current[i][1] === current[i][2] && current[i][0] === player)
      return true
    else if (current[0][i] === current[1][i] && current[1][i] === current[2][i] && current[0][i] === player)
      return true
  }

  // diagonals
  if (current[0][0] === current[1][1] && current[1][1] === current[2][2] && current[0][0] === player)
    return true

  if (current[2][0] === current[1][1] && current[1][1] === current[0][2] && current[2][0] === player)
    return true

  // No match.
  return false
}

/**
 * Check if a tic-tac-toe board is full
 * @param {Array[string]} board - Array of three strings of length three representing the board
 * @returns True if game board is full and false otherwise
 */
const isFull = (board) => {
  let full = true
  for(let i = 0; i < 3; i++)
  {
    if (board[i].indexOf(' ') >= 0)
    {
      full = false
      break;
    }
  }
  return full
}

const countPlayer = (board, player) => {
  let allSpaces = board[0] + board[1] + board[2]
  return allSpaces.split('').reduce((acc, current) => {
    if (current === player)
      return acc + 1
    return acc
  }, 0)
}

File: python/state-of-tic-tac-toe/state_of_tic_tac_toe.py
"""
Determine the state of a Tic-Tac-Toe board.
"""

def gamestate(board):
    """
    Determine the state of a tic-tac-toe board. Check if there is
    a winner, a draw or if it is ongoing. Raise an error if we are in an
    invalid state.
    :param board: A list of three strings of three characters each filled with X, O or a space for an empty spot
    :returns: 'win', 'draw' 'ongoing' depending on board state
    :raises: Value error if board is in invalid state. Doesn't detect invalid markers on the board.
    """
    winner_x = is_winner(board, 'X')
    winner_o = is_winner(board, 'O')

    if winner_x and winner_o:
        raise ValueError('Impossible board: game should have ended after the game was won')
    
    if winner_x or winner_o:
        return 'win'
    
    count_x = sum(1 for row in board for index in range(len(row)) if row[index] == 'X')
    count_o = sum(1 for row in board for index in range(len(row)) if row[index] == 'O')
    if count_x > count_o + 1:
        raise ValueError('Wrong turn order: X went twice')

    if count_o > count_x:
        raise ValueError('Wrong turn order: O started')
    
    if count_x + count_o < 9:
        return 'ongoing'
    return 'draw'

def is_winner(board, player):
    """
    Check to see if a player has won in tic-tac-toe
    :param board: A list of three strings of three characters each filled with X, O or a space for an empty spot
    :param player: The player to check. Should be 'X', or 'O'
    :returns: True if the winner.
    """
    win_row = player * 3
    # horizontal and vertical
    for index in range(3):
        if board[index] == win_row:
            return True
        vertical = board[0][index] + board[1][index] + board[2][index]
        if vertical == win_row:
            return True
    
    # diagonals
    left_to_right = board[0][0] + board[1][1] + board[2][2]
    right_to_left = board[0][2] + board[1][1] + board[2][0]
    if win_row in (left_to_right, right_to_left):
        return True
    
    return False